### PR TITLE
feat: support to new subscription service and make sense

### LINKED ISSUE #328: support to new subscription service and make sense
### Is your feature request related to a problem? Please describe.（如果该特性或者功能和已知某个问题相关，请描述问题是什么）

- New subscription service
- Make it work

### Describe the solution you'd like（问题的解决办法是什么）

- add one new subscription service
- design appropriate configuration items
- guaranteed a certain degree of reliability

### Describe alternatives you've considered（描述您考虑过的替代方案）

_No response_

### Additional context（其他补充内容）

_No response_

### PR DESCRIPTION
**Pull request by ZelinMa557**
### What problem does this PR solve?

Issue Number: close #328

#### Design Scheme:
* forward the write request in `/write` api
* init subscription writers when the server start
* use a goroutine to monitor the change of subscription
* use struct `SubscriberManager` to manage the subscription writers
* every subscription corresponds to a writer, each writer has a write request buffer, and has many goroutines to consume the buffer 
#### config:
* enabled
* http-timeout
* buffer-size
* concurrency
* insecure-skip-verify
* ca-certs
### How Has This Been Tested?
part of test result screen-shot:
image

image

image

image

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/config/subscriber.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	HTTPTimeout        int    `toml:"http-timeout"`
+	InsecureSkipVerify bool   `toml:"insecure-skip-verify"`
+	CaCerts            string `toml:"ca-certs"`
+	WriteBufferSize    int    `toml:"write-buffer-size"`
+	WriteConcurrency   int    `toml:"write-concurrency"`
+}
+
+func NewSubscriber() Subscriber {
+	return Subscriber{
+		Enabled:            true,
```
Comment (shilinlee) [inline lib/config/subscriber.go:38]:
```suggestion
		Enabled:            false,
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — APPROVED** [review]

File: coordinator/subscriber.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+	w.Send(wr)
+	atomic.StoreInt32(&w.i, (i+1)%int32(len(w.clients)))
+}
+
+type MetaClient interface {
+	Databases() map[string]*meta.DatabaseInfo
+	Database(string) (*meta.DatabaseInfo, error)
+}
+
+type SubscriberManger struct {
```
Comment (huang-feiteng) [inline coordinator/subscriber.go:198]: SubscriberManger -> SubscriberManager

File: app/ts-sql/sql/server.go [hunk 3]
```diff
@@ -223,6 +228,14 @@ func (s *Server) Open() error {
...
 	s.httpService.Handler.QueryExecutor.PointsWriter = s.PointsWriter
 	s.httpService.Handler.PointsWriter = s.PointsWriter
+	if s.SubscriberManger != nil {
+		s.httpService.Handler.SubscriberManger = s.SubscriberManger
+	}
+
+	if s.SubscriberManger != nil {
+		s.SubscriberManger.InitWriters()
+		go s.SubscriberManger.Update()
+	}
```
Comment (huang-feiteng) [inline app/ts-sql/sql/server.go:238]:
```go
	if s.SubscriberManger != nil {
		s.httpService.Handler.SubscriberManger = s.SubscriberManger
                s.SubscriberManger.InitWriters()
		go s.SubscriberManger.Update()
	}
```

**Review by huang-feiteng — CHANGES_REQUESTED** [review]

File: coordinator/subscriber.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
+/*
+Copyright 2022 Huawei Cloud Computing Technologies Co., Ltd.
```
Comment (ready-gogogo) [inline coordinator/subscriber.go:2]: Please change the copyright to 2023.

File: coordinator/subscriber.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+			for _, writer := range rp {
+				writer.Stop()
+			}
+		}
+	}
+}
+
+func (s *SubscriberManager) Update() {
+	for {
+		ch := s.client.(*metaclient.Client).WaitForDataChanged()
```
Comment (ready-gogogo) [inline coordinator/subscriber.go:362]: Is this conflict with the function waitForIndex? waitForIndex wait chis channel.

File: coordinator/subscriber.go (hunk 4 above)
Comment (ZelinMa557) [reply coordinator/subscriber.go:2]: ok

File: coordinator/subscriber.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+				writers = make([]SubscriberWriter, 0, len(rpi.Subscriptions))
+				changed = true
+			}
+			// record origin subscription names in a set
+			originSubs := make(map[string]bool)
+			for _, w := range writers {
+				originSubs[w.Name()] = true
+			}
+			// add new subscriptions
+			for _, sub := range rpi.Subscriptions {
```
[the remark this answers is not in the document]
Comment (ZelinMa557) [reply coordinator/subscriber.go:287]: The user cannot change the subscription directly, he or she must drop the subscription at first, then create a new subscription with the original name. It's inefficient to compare whether the subscription is changed or not.

File: coordinator/subscriber.go (hunk 5 above)
Comment (ZelinMa557) [reply coordinator/subscriber.go:362]: Actually there is no conflict, `pollForUpdates` will notify all the channels when metadata is changed, 
see https://github.com/openGemini/openGemini/blob/main/lib/metaclient/meta_client.go#L2399

File: open_src/influx/httpd/handler.go [hunk 7]
```diff
@@ -533,6 +551,16 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, user meta2.
 		h.Logger.Error("query error! parsing query value", zap.Error(err), zap.String("db", db), zap.Any("r", r))
 		return
 	}
+
+	// if subscription is not enabled,
+	// and the query contains subscription statements,
+	// then return error
+	if h.SubscriberManager == nil && containSubscriptionStatements(q.Statements) {
```
Comment (shilinlee) [inline open_src/influx/httpd/handler.go:558]: Doing this judgment on the query hot path will greatly reduce qps.

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: open_src/influx/httpd/handler.go (hunk 7 above)
Comment (ZelinMa557) [reply open_src/influx/httpd/handler.go:558]: > Doing this judgment on the query hot path will greatly reduce qps.
Yes, I need to use other method to disable the subscription statements when subscription is not enabled.

File: coordinator/subscriber.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+			changed := false
+			s.lock.RLock()
+			writers, ok := s.writers[dbi.Name][rpi.Name]
+			s.lock.RUnlock()
+			if !ok {
+				writers = make([]SubscriberWriter, 0, len(rpi.Subscriptions))
+				changed = true
+			}
+			// record origin subscription names in a set
+			originSubs := make(map[string]bool)
```
Comment (shilinlee) [inline coordinator/subscriber.go:282]:
```suggestion
			originSubs := make(map[string]struct{})
```

File: coordinator/subscriber.go [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+			for _, w := range writers {
+				originSubs[w.Name()] = true
+			}
+			// add new subscriptions
+			for _, sub := range rpi.Subscriptions {
+				if _, ok := originSubs[sub.Name]; !ok {
+					writer, err := s.NewSubscriberWriter(dbi.Name, rpi.Name, sub.Name, sub.Mode, sub.Destinations)
+					if err != nil {
+						s.Logger.Info("fail to create subscriber", zap.String("db", dbi.Name), zap.String("rp", rpi.Name), zap.String("sub", sub.Name),
```
Comment (shilinlee) [inline coordinator/subscriber.go:291]:
```suggestion
						s.Logger.Error("fail to create subscriber", zap.String("db", dbi.Name), zap.String("rp", rpi.Name), zap.String("sub", sub.Name), zap.Error(err)
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — APPROVED** [review]

File: config/openGemini.conf [hunk 10]
```diff
@@ -221,4 +221,12 @@
   # diff = 20
   # abs = 20000
   # max = 100000
-  # cool-down = "30m"
\ No newline at end of file
+  # cool-down = "30m"
+
+[subscriber]
+  # enabled = false
```
[the remark this answers is not in the document]
Comment (shilinlee) [reply config/openGemini.conf:250]: Keep `enable` item just like influx.

File: coordinator/subscriber.go [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+	}
+}
+
+type RoundRobinWriter struct {
+	BaseWriter
+	i int32
+}
+
+func (w *RoundRobinWriter) Write(lineProtocol []byte) {
+	i := atomic.LoadInt32(&w.i)
```
[the remark this answers is not in the document]
Comment (shilinlee) [reply coordinator/subscriber.go:186]: Check whether there is a concurrency problem here. If there is, it is recommended to `compareAndSwapInt32` first, and then `AddInt32`.

File: lib/config/subscriber.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import "errors"
+
+const (
+	DefaultHTTPTimeout      int = 30  // 30 seconds
```
Comment (shilinlee) [inline lib/config/subscriber.go:22]:
```suggestion
	DefaultHTTPTimeout      = 30 * time.Second  // 30 seconds
```

File: lib/config/subscriber.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	DefaultBufferSize       int = 100 // channel size 100
+	DefaultWriteConcurrency int = 15  // 15 goroutines
+)
+
+type Subscriber struct {
+	Enabled            bool   `toml:"enabled"`
+	HTTPTimeout        int    `toml:"http-timeout"`
```
Comment (shilinlee) [inline lib/config/subscriber.go:29]:
```suggestion
	HTTPTimeout        toml.Duration    `toml:"http-timeout"`
```

File: lib/config/subscriber.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+	InsecureSkipVerify bool   `toml:"insecure-skip-verify"`
+	HttpsCertificate   string `toml:"https-certificate"`
+	WriteBufferSize    int    `toml:"write-buffer-size"`
+	WriteConcurrency   int    `toml:"write-concurrency"`
+}
+
+func NewSubscriber() Subscriber {
+	return Subscriber{
+		Enabled:            false,
+		HTTPTimeout:        DefaultHTTPTimeout,
```
Comment (shilinlee) [inline lib/config/subscriber.go:39]:
```suggestion
		HTTPTimeout:        toml.Duration(DefaultHTTPTimeout(,
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: lib/config/subscriber.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+const (
+	DefaultHTTPTimeout = 30 * time.Second // 30 seconds
+	DefaultBufferSize  = 100              // channel size 100
+)
+
+type Subscriber struct {
+	Enabled            bool          `toml:"enabled"`
+	HTTPTimeout        time.Duration `toml:"http-timeout"`
```
Comment (shilinlee) [inline lib/config/subscriber.go:32]:
```suggestion
	HTTPTimeout        toml.Duration `toml:"http-timeout"`
```

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Review by shilinlee — APPROVED** [review]

**Review by scuzyj — APPROVED** [review]

**Pull request merged**